### PR TITLE
Introduce projection pushdown framework for JDBC connectors

### PR DESCRIPTION
**Pull request by Praveen2112**
## Description

This framework would allow us to pushdown specific functions closer to the JDBC datasource. We use this function to pushdown `reverse` function in case of PostgreSQL connector. In order to push down a specific function to the underlying JDBC datasource, we need two parts
 - Client should implement `convertProjection`
 - Write a `ProjectFunctionRule` specific to the function to be pushed down. 

We also support partial pushdown of expression i.e `LOWER(REVERSE(varchar_col))` will be translated into `LOWER(synthetic_col_from_query)` 

Spl thanks to @SemionPar for inspiring me to use `reverse` function for this framework. 

## Additional context and related issues

Without this change when we try to run a query like this 

```
explain select reverse(name) from nation
```
Current master
```
                                                               Query Plan
-----------------------------------------------------------------------------------------------------------------------------------------
 Trino version: testversion
 Fragment 0 [SOURCE]
     Output layout: [expr]
     Output partitioning: SINGLE []
     Output[columnNames = [_col0]]
     │   Layout: [expr:varchar(25)]
     │   Estimates: {rows: 25 (1.34kB), cpu: 0, memory: 0B, network: 0B}
     │   _col0 := expr
     └─ ScanProject[table = postgresql:tpch.nation tpch.nation columns=[name:varchar(25):varchar]]
            Layout: [expr:varchar(25)]
            Estimates: {rows: 25 (1.34kB), cpu: 1.34k, memory: 0B, network: 0B}/{rows: 25 (1.34kB), cpu: 1.34k, memory: 0B, network: 0B}
            expr := reverse(name)
            name := name:varchar(25):varchar
```

With this optimization

```
trino:tpch> explain select reverse(name) from nation;
                                                                    Query Plan
--------------------------------------------------------------------------------------------------------------------------------------------------
 Trino version: testversion
 Fragment 0 [SOURCE]
     Output layout: [pfgnrtd]
     Output partitioning: SINGLE []
     Output[columnNames = [_col0]]
     │   Layout: [pfgnrtd:varchar(25)]
     │   Estimates: {rows: ? (?), cpu: 0, memory: 0B, network: 0B}
     │   _col0 := pfgnrtd
     └─ TableScan[table = postgresql:Query[SELECT REVERSE("name") AS "_pfgnrtd_0" FROM "tpch"."nation"] columns=[_pfgnrtd_0:varchar(25):varchar]]
            Layout: [pfgnrtd:varchar(25)]
            Estimates: {rows: ? (?), cpu: ?, memory: 0B, network: 0B}
            pfgnrtd := _pfgnrtd_0:varchar(25):varchar
```

## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
(x) Release notes are required. Please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# PostgreSQL
* Pushdown `reverse` function to the datasource when applied as a projection expression 
```

### CODE REVIEW

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/DefaultJdbcMetadata.java [hunk 1]
```diff
@@ -350,6 +362,151 @@ public Optional<ProjectionApplicationResult<ConnectorTableHandle>> applyProjecti
...
+                    session,
+                    nextSyntheticColumnId,
+                    child,
+                    assignments,
+                    translatedExpression);
+            nextSyntheticColumnId = rewrittenExpression.nextSyntheticColumnId;
+            columnsBuilder.addAll(rewrittenExpression.columnHandles);
+            assignmentBuilder.addAll(rewrittenExpression.assignments);
+            syntheticVariablesBuilder.putAll(rewrittenExpression.syntheticVariables);
+            columnExpressionsBuilder.putAll(rewrittenExpression.columnExpressions);
```
Comment (Praveen2112) [inline plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/DefaultJdbcMetadata.java:483]: Too many objects might be created - I'm not sure if we could pass builder as arguments here.

**Comment by huw0** [comment]
This PR is quite timely as I was looking at the possibility to do something similar today. Thanks @Praveen2112.

My primary use case is pushdown of JSON field extraction which causes a significant speedup to the overall query.

However ideally it'd be good to enable pushdown of all functions that the connector source supports where the expression matches that of Trino's own.

It'd be really helpful if reuse of function pushdown could be easily reusable between `convertPredicate` and `convertProjection`?
This adds some complexity so may be better as a separate PR but I wonder if it is worth early consideration.

File: plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/rule/RewriteStringReverseFunction.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+import static io.trino.plugin.base.expression.ConnectorExpressionPatterns.argument;
+import static io.trino.plugin.base.expression.ConnectorExpressionPatterns.argumentCount;
+import static io.trino.plugin.base.expression.ConnectorExpressionPatterns.call;
+import static io.trino.plugin.base.expression.ConnectorExpressionPatterns.functionName;
+import static io.trino.plugin.base.expression.ConnectorExpressionPatterns.type;
+import static io.trino.plugin.base.expression.ConnectorExpressionPatterns.variable;
+
+public class RewriteStringReverseFunction
+        implements ProjectFunctionRule<JdbcExpression, ParameterizedExpression>
+
```
Comment (ssheikin) [inline plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/rule/RewriteStringReverseFunction.java:42]: unnecessary new line

File: plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/rule/RewriteStringReverseFunction.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+import static io.trino.matching.Capture.newCapture;
+import static io.trino.plugin.base.expression.ConnectorExpressionPatterns.argument;
+import static io.trino.plugin.base.expression.ConnectorExpressionPatterns.argumentCount;
+import static io.trino.plugin.base.expression.ConnectorExpressionPatterns.call;
+import static io.trino.plugin.base.expression.ConnectorExpressionPatterns.functionName;
+import static io.trino.plugin.base.expression.ConnectorExpressionPatterns.type;
+import static io.trino.plugin.base.expression.ConnectorExpressionPatterns.variable;
+
+public class RewriteStringReverseFunction
```
Comment (ssheikin) [inline plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/rule/RewriteStringReverseFunction.java:40]: Should it be common for jdbc connectors, not only for PostgreSql exclusively?

File: plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/rule/RewriteStringReverseFunction.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        implements ProjectFunctionRule<JdbcExpression, ParameterizedExpression>
```
Comment (ssheikin) [inline plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/rule/RewriteStringReverseFunction.java:41]: nit: Could you please extract implementations of the rule to a separate commit?
This is ok to have it in the same commit, but having in separate will ease to figure out a scope of the change.

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/DefaultJdbcMetadata.java [hunk 5]
```diff
@@ -350,6 +362,151 @@ public Optional<ProjectionApplicationResult<ConnectorTableHandle>> applyProjecti
...
+    {
+        int nextSyntheticColumnId = handle.getNextSyntheticColumnId();
+
+        ImmutableSet.Builder<JdbcColumnHandle> newColumnsBuilder = ImmutableSet.builder();
+        ImmutableList.Builder<Assignment> assignmentBuilder = ImmutableList.builder();
+        ImmutableMap.Builder<ConnectorExpression, Variable> newVariablesBuilder = ImmutableMap.builder();
+        ImmutableMap.Builder<String, ParameterizedExpression> columnExpressionsBuilder = ImmutableMap.builder();
+        Set<ConnectorExpression> translatedExpression = new HashSet<>();
+
+        if (isComplexExpressionPushdown(session)) {
```
Comment (ssheikin) [inline plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/DefaultJdbcMetadata.java:379]: This could be a first statement in the method, and if false - `return Optional.empty();`

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/DefaultJdbcMetadata.java [hunk 6]
```diff
@@ -350,6 +362,151 @@ public Optional<ProjectionApplicationResult<ConnectorTableHandle>> applyProjecti
...
+            for (ConnectorExpression projection : projections) {
+                RewrittenExpression rewrittenExpression = rewriteExpression(session, nextSyntheticColumnId, projection, assignments, translatedExpression);
+                nextSyntheticColumnId = rewrittenExpression.nextSyntheticColumnId;
+                newColumnsBuilder.addAll(rewrittenExpression.columnHandles);
+                assignmentBuilder.addAll(rewrittenExpression.assignments);
+                newVariablesBuilder.putAll(rewrittenExpression.syntheticVariables);
+                columnExpressionsBuilder.putAll(rewrittenExpression.columnExpressions);
```
Comment (ssheikin) [inline plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/DefaultJdbcMetadata.java:386]: no need in this assignments and builders. Just use rewrittenExpression

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/DefaultJdbcMetadata.java [hunk 7]
```diff
@@ -350,6 +362,151 @@ public Optional<ProjectionApplicationResult<ConnectorTableHandle>> applyProjecti
...
+
+        // Modify projections to refer to new variables
+        Map<ConnectorExpression, Variable> newVariables = newVariablesBuilder.buildOrThrow();
+        List<ConnectorExpression> newProjections = projections.stream()
+                .map(expression -> replaceWithNewVariables(expression, newVariables))
+                .collect(toImmutableList());
+
+        List<Assignment> outputAssignments = assignmentBuilder.build();
+
+        if (!newVariables.isEmpty()) {
```
Comment (ssheikin) [inline plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/DefaultJdbcMetadata.java:400]: this will disappear after 
https://github.com/trinodb/trino/pull/22203/files#r1630996112

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/DefaultJdbcMetadata.java [hunk 8]
```diff
@@ -350,6 +362,151 @@ public Optional<ProjectionApplicationResult<ConnectorTableHandle>> applyProjecti
...
+            List<ConnectorExpression> projections,
+            Map<String, ColumnHandle> assignments)
+    {
+        int nextSyntheticColumnId = handle.getNextSyntheticColumnId();
+
+        ImmutableSet.Builder<JdbcColumnHandle> newColumnsBuilder = ImmutableSet.builder();
+        ImmutableList.Builder<Assignment> assignmentBuilder = ImmutableList.builder();
+        ImmutableMap.Builder<ConnectorExpression, Variable> newVariablesBuilder = ImmutableMap.builder();
+        ImmutableMap.Builder<String, ParameterizedExpression> columnExpressionsBuilder = ImmutableMap.builder();
+        Set<ConnectorExpression> translatedExpression = new HashSet<>();
```
Comment (ssheikin) [inline plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/DefaultJdbcMetadata.java:377]: I didn't grasp the PR fully, but for now it looks like this variable could have a little bit more ugly name, to gather more attention. E.g. `translatedExpressionAccumulator`

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/DefaultJdbcMetadata.java [hunk 9]
```diff
@@ -350,6 +362,151 @@ public Optional<ProjectionApplicationResult<ConnectorTableHandle>> applyProjecti
...
+            return Optional.of(new ProjectionApplicationResult<>(
+                    new JdbcTableHandle(
+                            new JdbcQueryRelationHandle(preparedQuery),
+                            TupleDomain.all(),
+                            ImmutableList.of(),
+                            Optional.empty(),
+                            OptionalLong.empty(),
+                            Optional.of(newColumns),
+                            handle.getOtherReferencedTables(),
+                            nextSyntheticColumnId,
```
Comment (ssheikin) [inline plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/DefaultJdbcMetadata.java:411]: Could it be already used by
`RewrittenExpression rewrittenExpression = rewriteExpression(session, nextSyntheticColumnId, projection, assignments, translatedExpression);` ?

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/DefaultJdbcMetadata.java [hunk 10]
```diff
@@ -350,6 +362,151 @@ public Optional<ProjectionApplicationResult<ConnectorTableHandle>> applyProjecti
...
+            columnExpressionsBuilder.putAll(rewrittenExpression.columnExpressions);
+        }
+
+        return new RewrittenExpression(
+                nextSyntheticColumnId,
+                syntheticVariablesBuilder.buildOrThrow(),
+                columnExpressionsBuilder.buildOrThrow(),
+                columnsBuilder.build(),
+                assignmentBuilder.build());
+    }
```
Comment (ssheikin) [inline plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/DefaultJdbcMetadata.java:492]: Looks like it can be extacted to a separate commit

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/DefaultJdbcMetadata.java [hunk 11]
```diff
@@ -350,6 +362,151 @@ public Optional<ProjectionApplicationResult<ConnectorTableHandle>> applyProjecti
...
+
+    public record RewrittenExpression(
```
Comment (ssheikin) [inline plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/DefaultJdbcMetadata.java:494]: Should we move it to the very bottom of the class?

File: lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/projection/ProjectFunctionRule.java [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+import io.trino.spi.connector.ConnectorSession;
+import io.trino.spi.expression.ConnectorExpression;
+
+import java.util.Map;
+import java.util.Optional;
+
+import static com.google.common.base.Verify.verifyNotNull;
+import static java.util.Objects.requireNonNull;
+
+public interface ProjectFunctionRule<ProjectionResult, ExpressionResult>
```
Comment (ssheikin) [inline lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/projection/ProjectFunctionRule.java:28]: looks like `ProjectFunctionRule` and `AggregateFunctionRule` have many in common. 
Should they share some parent?

**Comment by ssheikin** [comment]
skimmed

File: plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/rule/RewriteStringReverseFunction.java (hunk 3 above)
Comment (Praveen2112) [reply plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/rule/RewriteStringReverseFunction.java:40]: For now - I think it could be exclusive to PostgreSql and we could migrate them to other connector if required.

File: plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/rule/RewriteStringReverseFunction.java (hunk 4 above)
Comment (Praveen2112) [reply plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/rule/RewriteStringReverseFunction.java:41]: We could but - moving it outside this commit - makes this commit un-tested righti.e the framework is untested without this rule.

File: lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/projection/ProjectFunctionRule.java (hunk 12 above)
Comment (Praveen2112) [reply lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/projection/ProjectFunctionRule.java:28]: For now - I think it is better to have them as independent as possible - having some common parent would allow some rule to be creeped across the framework.

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/DefaultJdbcMetadata.java (hunk 5 above)
Comment (Praveen2112) [reply plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/DefaultJdbcMetadata.java:379]: I don't think we need to return `Optional#empty` as we need to additional changes for normal expression.

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/DefaultJdbcMetadata.java (hunk 9 above)
Comment (Praveen2112) [reply plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/DefaultJdbcMetadata.java:411]: I didn't get this fully.

**Comment by Praveen2112** [comment]
@ssheikin Thanks a lot for the review. AC

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/DefaultJdbcMetadata.java (hunk 5 above)
Comment (ssheikin) [reply plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/DefaultJdbcMetadata.java:387]: `if (!newVariables.isEmpty()) {` is true only if `newVariables` is populated. 
`newVariables` is populated only if `newVariablesBuilder` is populated.
`newVariablesBuilder` is populated only when `if (isComplexExpressionPushdown(session)) {` is true.
What do I miss?

Comment (Praveen2112) [reply plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/DefaultJdbcMetadata.java:387]: Okay !! I get it - Previously I was a bit confused as this code snippet was a part of `applyProjection`. Thanks for pointing it out.

**Comment by ebyhr** [comment]
The change may cause regression when `reverse` function pushdown happens on non-varchar types (e.g. `money` type) in PostgreSQL.

**Comment by Praveen2112** [comment]
The same can be restricted via rules right ? `Reverse` pushdown rule can be applied only on `JdbcTypeHandle` which corresponds to `VARCHAR` - Would it solve the issue ?

**Comment by Praveen2112** [comment]
@ebyhr Thank you for pointing the regression. Have fixed it and also added a test case to verify it doesn't break for columns with special data type.

File: plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlConnectorTest.java [hunk 13]
```diff
@@ -1026,6 +1042,120 @@ public void testTimestampColumnAndTimestampWithTimeZoneConstant()
...
+                    .matches("VALUES 'cba'")
+                    .isFullyPushedDown();
+
+            assertThat(query("SELECT id, reverse(reverse(reverse(reverse(reverse(varchar_col))))) FROM " + table.getName()))
+                    .skippingTypesCheck()
+                    .matches("VALUES (BIGINT '1', 'cba'), (BIGINT '2', NULL)")
+                    .isNotFullyPushedDown(ProjectNode.class)
+                    .hasPlan(output(
+                            project(
+                                    tableScan(table.getName(), ImmutableMap.of("varchar_col", "varchar_col")))));
```
[the remark this answers is not in the document]
Comment (ssheikin) [reply plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlConnectorTest.java]: I see, it's next commit. Nice.

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/DefaultJdbcMetadata.java [hunk 14]
```diff
@@ -358,6 +372,132 @@ public Optional<ProjectionApplicationResult<ConnectorTableHandle>> applyProjecti
...
+            assignmentBuilder.addAll(rewrittenExpression.assignments());
+            syntheticVariablesBuilder.putAll(rewrittenExpression.syntheticVariables());
+            columnExpressionsBuilder.putAll(rewrittenExpression.columnExpressions());
+        }
+        return new RewrittenExpression(
+                nextSyntheticColumnId,
+                syntheticVariablesBuilder.buildOrThrow(),
+                columnExpressionsBuilder.buildOrThrow(),
+                columnsBuilder.build(),
+                assignmentBuilder.build());
```
Comment (ssheikin) [inline plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/DefaultJdbcMetadata.java:498]: is it possible / does it make sense to keep order of parameters / `addAll` the same?

**Review by ssheikin — APPROVED** [review]
lgtm

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/DefaultJdbcMetadata.java [hunk 15]
```diff
@@ -358,6 +371,101 @@ public Optional<ProjectionApplicationResult<ConnectorTableHandle>> applyProjecti
...
+            Map<String, ColumnHandle> assignments)
+    {
+        int nextSyntheticColumnId = handle.getNextSyntheticColumnId();
+
+        ImmutableSet.Builder<JdbcColumnHandle> newColumnsBuilder = ImmutableSet.builder();
+        ImmutableList.Builder<Assignment> assignmentBuilder = ImmutableList.builder();
+        ImmutableMap.Builder<ConnectorExpression, Variable> newVariablesBuilder = ImmutableMap.builder();
+        ImmutableMap.Builder<String, ParameterizedExpression> columnExpressionsBuilder = ImmutableMap.builder();
+
+        for (ConnectorExpression projection : ImmutableSet.copyOf(projections)) {
```
[the remark this answers is not in the document]
Comment (Praveen2112) [reply plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/DefaultJdbcMetadata.java:389]: This would allow to skip processing redundant expression - if propagated

**Comment by findinpath** [comment]
@Praveen2112  would you mind adding more hands-on details on the description of the PR about the problem that you're solving?

Also helpful would be potentially to share an EXPLAIN plan of a query with and without your changes.

**Comment by Praveen2112** [comment]
@findinpath Yeah sure !! Would add the `EXPLAIN` set of changes -

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/DefaultJdbcMetadata.java [hunk 16]
```diff
@@ -358,6 +371,101 @@ public Optional<ProjectionApplicationResult<ConnectorTableHandle>> applyProjecti
...
+
+        // Modify projections to refer to new variables
+        Map<ConnectorExpression, Variable> newVariables = newVariablesBuilder.buildOrThrow();
+        List<ConnectorExpression> newProjections = projections.stream()
+                .map(expression -> replaceWithNewVariables(expression, newVariables))
+                .collect(toImmutableList());
+
+        List<Assignment> outputAssignments = assignmentBuilder.build();
+
+        if (newVariables.isEmpty()) {
```
Comment (ssheikin) [inline plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/DefaultJdbcMetadata.java:408]: may be moved even upper.

**Comment by Praveen2112** [comment]
@ssheikin Thanks for the review. AC

**Review by ssheikin — APPROVED** [review]

File: plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/rule/RewriteStringReverseFunction.java [hunk 17]
```diff
@@ -0,0 +1,77 @@
...
+        }
+        Optional<ParameterizedExpression> translatedArgument = context.rewriteExpression(argument);
+        if (translatedArgument.isEmpty()) {
+            return Optional.empty();
+        }
+
+        return Optional.of(new JdbcExpression(
+                "REVERSE(%s)".formatted(translatedArgument.get().expression()),
+                ImmutableList.copyOf(translatedArgument.get().parameters()),
+                ((JdbcColumnHandle) context.getAssignment(argument.getName())).getJdbcTypeHandle()));
```
Comment (ebyhr) [inline plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/rule/RewriteStringReverseFunction.java:75]: Can we replace `((JdbcColumnHandle) context.getAssignment(argument.getName()))` with `columnHandle`?

File: plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/rule/RewriteStringReverseFunction.java [hunk 18]
```diff
@@ -0,0 +1,77 @@
...
+
+    @Override
+    public Optional<JdbcExpression> rewrite(ConnectorExpression projectionExpression, Captures captures, RewriteContext<ParameterizedExpression> context)
+    {
+        Variable argument = captures.get(ARGUMENT);
+        JdbcColumnHandle columnHandle = (JdbcColumnHandle) context.getAssignment(argument.getName());
+        // Skip rewriting the expression for special data types which are treated as Varchar
+        if (JDBCType.valueOf(columnHandle.getJdbcTypeHandle().jdbcType()) != JDBCType.VARCHAR) {
+            return Optional.empty();
+        }
```
Comment (ebyhr) [inline plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/rule/RewriteStringReverseFunction.java:66]: This protection looks still insufficient. PostgreSQL connector may return `JDBCType.VARCHAR` on enum type that doesn't support `reverse` function. Probably, we need to check `JdbcTypeHandle#jdbcTypeName` like:
https://github.com/trinodb/trino/blob/452c627fe7e91bb21e7f1ecb591843c27bc4a673/plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java#L610-L615

**Comment by Praveen2112** [comment]
@ebyhr Thanks a lot for the review - AC

**Review by ebyhr — APPROVED** [review]
